crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+            let observed = ProviderJWKs {
+                issuer: issuer.clone(),
+                version: state.on_chain_version() + 1,
+                jwks,
+            };
+            let signature = self
+                .consensus_key
+                .sign(&observed)
+                .map_err(|e| anyhow!("crypto material error occurred duing signing: {}", e))?;
+            let abort_handle = self.certified_update_producer.start_produce(
+                self.epoch_state.clone(),
+                observed.clone(),
+                self.qc_update_tx.clone(),
+            );
+            state.consensus_state = ConsensusState::InProgress {
+                my_proposal: ObservedUpdate {
+                    author: self.my_addr,
+                    observed: observed.clone(),
+                    signature,
+                },
+                abort_handle_wrapper: QuorumCertProcessGuard::new(abort_handle),
+            };
+            info!("[JWK] update observed, update={:?}", observed);
+        }
+
+        Ok(())
+    }
+
+    /// Invoked on start, or on on-chain JWK updated event.
+    pub fn reset_with_on_chain_state(&mut self, on_chain_state: AllProvidersJWKs) -> Result<()> {
+        debug!(
+            "[JWK] reset_with_on_chain_state: BEGIN: on_chain_state={:?}",
+            on_chain_state
+        );
+        let onchain_issuer_set: HashSet<Issuer> = on_chain_state
+            .entries
+            .iter()
+            .map(|entry| entry.issuer.clone())
+            .collect();
+        self.states_by_issuer
+            .retain(|issuer, _| onchain_issuer_set.contains(issuer));
+        for provider_jwks in on_chain_state.entries {
+            let x = self
+                .states_by_issuer
+                .get(&provider_jwks.issuer)
+                .and_then(|s| s.on_chain.as_ref());
+            if x != Some(&provider_jwks) {
+                self.states_by_issuer.insert(
+                    provider_jwks.issuer.clone(),
+                    PerProviderState::new(provider_jwks),
+                );
+            }
+        }
+        Ok(())
+    }
+
+    pub fn process_peer_request(&mut self, rpc_req: IncomingRpcRequest) -> Result<()> {
+        let IncomingRpcRequest {
+            msg,
+            mut response_sender,
+            sender,
+        } = rpc_req;
+        debug!(
+            "[JWK] process_peer_request: sender={}, is_self={}",
+            sender,
+            sender == self.my_addr
+        );
+        match msg {
+            JWKConsensusMsg::ObservationRequest(request) => {
+                let state = self.states_by_issuer.entry(request.issuer).or_default();
+                let response: Result<JWKConsensusMsg> = match &state.consensus_state {
+                    ConsensusState::NotStarted => Err(anyhow!("observed update unavailable")),
+                    ConsensusState::InProgress { my_proposal, .. }
+                    | ConsensusState::Finished { my_proposal, .. } => Ok(
```

Comment:
reset_with_onchain_state doesn't update consensus state? I thought we had states like
```
OnChain(..),
QuorumInProgress(..),
TransactionInProgress(..),
```

Reply:
The latest plan is: if the on-chain update doesn't modify issuer A's key, we don't have to reset issuer A's local state. This way, in case there's an in-progress certifying session for A, it won't be killed.

(is this comment placed on the wrong line?)